heft-plugins/heft-lint-plugin/src/SarifFormatter.ts

```diff
@@ -0,0 +1,258 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import url from 'url';
+
+import type * as TEslint from 'eslint';
+
+export interface ISerifFormatterOptions {
+  ignoreSuppressed: boolean;
+  eslintVersion?: string;
+}
+
+interface ISarifRun {
+  tool: {
+    driver: {
+      name: string;
+      informationUri: string;
+      version?: string;
+      rules: IStaticAnalysisRules[];
+    };
+  };
+  artifacts?: string[];
+  results?: ISarifRepresentation[];
+  invocations?: {
+    toolConfigurationNotifications: ISarifRepresentation[];
+    executionSuccessful: boolean;
+  }[];
+}
+
+interface ISarifFile {
+  location: {
+    uri: string;
+  };
+}
+
+interface ISarifRepresentation {
+  level: string;
+  message: {
+    text: string;
+  };
+  locations: ISarifLocation[];
+  ruleId?: string;
+  descriptor?: {
+    id: string;
+  };
+  suppressions?: ISuppressedAnalysis[];
+}
+
+interface ISuppressedAnalysis {
+  kind: string;
+  justification: string;
+}
+
+// Interface for the SARIF log structure
+interface ISarifLog {
+  version: string;
+  $schema: string;
+  runs: ISarifRun[];
+}
+
+interface IRegion {
+  startLine?: number;
+  startColumn?: number;
+  endLine?: number;
+  endColumn?: number;
+  snippet?: {
+    text: string;
+  };
+}
+
+interface IStaticAnalysisRules {
+  id: string;
+  name?: string;
+  shortDescription?: {
+    text: string;
+  };
+  fullDescription?: {
+    text: string;
+  };
+  defaultConfiguration?: {
+    level: 'note' | 'warning' | 'error';
+  };
+  helpUri?: string;
+  properties?: {
+    category?: string;
+    precision?: 'very-high' | 'high' | 'medium' | 'low';
+    tags?: string[];
+    problem?: {
+      severity?: 'recommendation' | 'warning' | 'error';
+      securitySeverity?: number;
+    };
+  };
+}
+
+interface ISarifLocation {
+  physicalLocation: ISarifPhysicalLocation;
+}
+
+interface ISarifPhysicalLocation {
+  artifactLocation: {
+    uri: string;
+    index: number;
+  };
+  region?: IRegion;
+}
+
+interface IMessage extends TEslint.Linter.LintMessage {
+  suppressions?: ISuppressedAnalysis[];
+}
+
+interface IExtendedLintResult extends TEslint.ESLint.LintResult {
+  suppressedMessages: TEslint.ESLint.LintResult['suppressedMessages'];
+}
+
+const sarifFiles: Map<string, ISarifFile> = new Map();
+const sarifResults: ISarifRepresentation[] = [];
+
+const internalErrorId: string = 'ESL0999';
+const toolConfigurationNotifications: ISarifRepresentation[] = [];
+
+const sarifRun: ISarifRun = {
+  tool: {
```

Comment:
This shouldn't be a module level const. Anything that gets modified by the functions in this module should be local to the function. Otherwise you preserve the values you set in the functions between calls